neqo-transport/src/connection/mod.rs

```diff
@@ -3074,14 +3078,23 @@ impl Connection {
         }
     }
 
-    fn decode_ack_delay(&self, v: u64) -> Duration {
+    fn decode_ack_delay(&self, v: u64) -> Res<Duration> {
         // If we have remote transport parameters, use them.
         // Otherwise, ack delay should be zero (because it's the handshake).
         self.tps.borrow().remote.as_ref().map_or_else(
-            || Duration::new(0, 0),
+            || Ok(Duration::default()),
             |r| {
-                let exponent = u32::try_from(r.get_integer(tparams::ACK_DELAY_EXPONENT)).unwrap();
-                Duration::from_micros(v.checked_shl(exponent).unwrap_or(u64::MAX))
+                let exponent = u32::try_from(r.get_integer(tparams::ACK_DELAY_EXPONENT))?;
+                if exponent > 20 {
+                    // ACK_DELAY_EXPONENT > 20 is invalid per RFC9000
+                    return Err(Error::TransportParameterError);
+                }
+                let corrected = if v.leading_zeros() >= exponent {
+                    v << exponent
+                } else {
+                    u64::MAX
+                };
```

Comment:
This seems good.  Should we have a saturating shift function?  This isn't the first time we've needed this sort of thing.

Reply:
Maybe long-term we can use:

https://doc.rust-lang.org/beta/std/num/struct.Saturating.html#method.rotate_left

Reply:
New issue: https://github.com/mozilla/neqo/issues/2402